### PR TITLE
keepalived: T4109: Add high-availability virtual-server

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Add new feature:
```
set high-availability virtual-server
```
Move vrrp scripts/templates/xml under node `high-availability`
As `high-availability.py` is handler for both `vrrp` and `virtual-server`
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4109

## Component(s) name

keepalive, virtual-server
## How to test
```
set high-availability virtual-server 203.0.113.111 algorithm 'least-connection'
set high-availability virtual-server 203.0.113.111 delay-loop '10'
set high-availability virtual-server 203.0.113.111 forward-method 'nat'
set high-availability virtual-server 203.0.113.111 persistence-timeout '600'
set high-availability virtual-server 203.0.113.111 port '2222'
set high-availability virtual-server 203.0.113.111 protocol 'tcp'
set high-availability virtual-server 203.0.113.111 real-server 192.0.2.2 connection-timeout '30'
set high-availability virtual-server 203.0.113.111 real-server 192.0.2.2 port '22'
set high-availability virtual-server 203.0.113.111 real-server 192.0.2.12 connection-timeout '30'
set high-availability virtual-server 203.0.113.111 real-server 192.0.2.12 port '22'
set high-availability virtual-server 203.0.113.111 real-server 192.0.2.14 connection-timeout '30'
set high-availability virtual-server 203.0.113.111 real-server 192.0.2.14 port '22'
```

Keepalived configuration:
```
vyos@r11-roll# cat /run/keepalived/keepalived.conf 
# Autogenerated by VyOS
# Do not edit this file, all your changes will be lost
# on next commit or reboot

global_defs {
    dynamic_interfaces
    script_user root
    notify_fifo /run/keepalived/keepalived_notify_fifo
    notify_fifo_script /usr/libexec/vyos/system/keepalived-fifo.py
}



# Virtual-server configuration
virtual_server 203.0.113.111 2222 {
    delay_loop 10
    lb_algo lc
    lb_kind NAT
    persistence_timeout 600
    protocol TCP
    real_server 192.0.2.2 22 {
        weight 1
        TCP_CHECK {
            connect_timeout 30
        }
    }
    real_server 192.0.2.12 22 {
        weight 1
        TCP_CHECK {
            connect_timeout 30
        }
    }
    real_server 192.0.2.14 22 {
        weight 1
        TCP_CHECK {
            connect_timeout 30
        }
    }
}

```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by c-po** [comment]
Looks good to me. Can you please extend the smoketests to also cover the load-balancing topic?

**Comment by sever-sever** [comment]
> Looks good to me. Can you please extend the smoketests to also cover the load-balancing topic?

@c-po Done

One small notice. 
`Last Transition` for vrrp is changed after adding ``virtual-server` for the current vrrp config
```
vyos@r11-roll# run show vrrp 
Name    Interface      VRID  State      Priority  Last Transition
------  -----------  ------  -------  ----------  -----------------
GRP01   eth1             10  MASTER          150  7s
GRP02   eth0             20  MASTER          150  7s
```

But MASTER state wasn't changed, so I think it is OK.

**Review by c-po — APPROVED** [review]

File: smoketest/scripts/cli/test_ha_virtual_server.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+from vyos.template import inc_ip
+
+PROCESS_NAME = 'keepalived'
+KEEPALIVED_CONF = VRRP.location['config']
+base_path = ['high-availability']
+
+vrrp_interface = 'eth1'
+
+def getConfig(string, end='}'):
+    command = f'cat {KEEPALIVED_CONF} | sed -n "/^{string}/,/^{end}/p"'
```
Comment (dmbaturin) [inline smoketest/scripts/cli/test_ha_virtual_server.py]: Why are you shelling out here? That would be just a few lines to do in Python.

File: interface-definitions/high-availability.xml.in [hunk 2]
```diff
@@ -252,6 +252,144 @@
...
+              </valueHelp>
+              <constraint>
+                <validator name="numeric" argument="--range 1-3600"/>
+              </constraint>
+            </properties>
+            <defaultValue>10</defaultValue>
+          </leafNode>
+          <leafNode name="forward-method">
+            <properties>
+              <help>Forwarding method (default - NAT)</help>
```
Comment (dmbaturin) [inline interface-definitions/high-availability.xml.in]: In this file, there are _three_ different formatting variants of the detault value: with a hyphen, with a colon, and without a separator. ;)

Let's standardize on `(default: $default) like on line 305.

File: interface-definitions/high-availability.xml.in [hunk 3]
```diff
@@ -252,6 +252,144 @@
...
+            </properties>
+            <defaultValue>nat</defaultValue>
+          </leafNode>
+          #include <include/port-number.xml.i>
+          <leafNode name="persistence-timeout">
+            <properties>
+              <help>Timeout for persistent connections</help>
+              <valueHelp>
+                <format>u32:1-86400</format>
+                <description>Timeout for persistent connections (default 300)</description>
```
Comment (dmbaturin) [inline interface-definitions/high-availability.xml.in]: Let's add a colon here before 300.

File: interface-definitions/high-availability.xml.in [hunk 4]
```diff
@@ -252,6 +252,144 @@
...
+              </valueHelp>
+              <constraint>
+                <validator name="numeric" argument="--range 1-86400"/>
+              </constraint>
+            </properties>
+            <defaultValue>300</defaultValue>
+          </leafNode>
+          <leafNode name="protocol">
+            <properties>
+              <help>Protocol for port checks (default TCP)</help>
```
Comment (dmbaturin) [inline interface-definitions/high-availability.xml.in]: Another case of inconsistent default value help formatting.

File: interface-definitions/high-availability.xml.in [hunk 5]
```diff
@@ -252,6 +252,144 @@
...
+              <completionHelp>
+                <list>tcp udp</list>
+              </completionHelp>
+              <valueHelp>
+                <format>tcp</format>
+                <description>Protocol TCP</description>
```
Comment (dmbaturin) [inline interface-definitions/high-availability.xml.in]: Do we need the "protocol" word here? People generally know what TCP is.

File: interface-definitions/high-availability.xml.in [hunk 6]
```diff
@@ -252,6 +252,144 @@
...
+          </leafNode>
+          <tagNode name="real-server">
+            <properties>
+              <help>Real server address</help>
+            </properties>
+            <children>
+              #include <include/port-number.xml.i>
+              <leafNode name="connection-timeout">
+                <properties>
+                  <help>Connection timeout to remote server</help>
```
Comment (dmbaturin) [inline interface-definitions/high-availability.xml.in]: I'd reword to "Server connection timeout".

**Review by dmbaturin — CHANGES_REQUESTED** [review]
Looks good overall, but I made a few suggestions.

File: interface-definitions/high-availability.xml.in (hunk 2 above)
Comment (sever-sever) [reply interface-definitions/high-availability.xml.in]: Done

File: interface-definitions/high-availability.xml.in (hunk 3 above)
Comment (sever-sever) [reply interface-definitions/high-availability.xml.in]: Done

File: interface-definitions/high-availability.xml.in (hunk 4 above)
Comment (sever-sever) [reply interface-definitions/high-availability.xml.in]: Done

File: interface-definitions/high-availability.xml.in (hunk 5 above)
Comment (sever-sever) [reply interface-definitions/high-availability.xml.in]: Fixed

File: interface-definitions/high-availability.xml.in (hunk 6 above)
Comment (sever-sever) [reply interface-definitions/high-availability.xml.in]: Done

File: smoketest/scripts/cli/test_ha_virtual_server.py (hunk 1 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_ha_virtual_server.py]: Replaced with `with open(KEEPALIVED_CONF) as f:`

Comment (c-po) [reply smoketest/scripts/cli/test_ha_virtual_server.py]: Why not use `vyos.util.read_file()`?

Comment (sever-sever) [reply smoketest/scripts/cli/test_ha_virtual_server.py]: > Why not use `vyos.util.read_file()`?

It depends on what we want to read.
Do we want to read a whole config file or just a section in the config file which include configuration related virtual_server
Both options work for me

**Comment by c-po** [comment]
> > Looks good to me. Can you please extend the smoketests to also cover the load-balancing topic?
> 
> @c-po Done
> 
> One small notice. `Last Transition` for vrrp is changed after adding ``virtual-server` for the current vrrp config
> 
> ```
> vyos@r11-roll# run show vrrp 
> Name    Interface      VRID  State      Priority  Last Transition
> ------  -----------  ------  -------  ----------  -----------------
> GRP01   eth1             10  MASTER          150  7s
> GRP02   eth0             20  MASTER          150  7s
> ```
> 
> But MASTER state wasn't changed, so I think it is OK.

That happens because a systemvtl reload is sent to keepalived re-reading the configuration.

File: smoketest/scripts/cli/test_ha_virtual_server.py (hunk 1 above)
Comment (c-po) [reply smoketest/scripts/cli/test_ha_virtual_server.py]: with open and read_file do exactly the same. The latter is only a wrapper where we could inject debug code and/or assime default behavior.

File: src/conf_mode/high-availability.py [hunk 7]
```diff
@@ -40,33 +40,41 @@ def get_config(config=None):
...
-def verify(vrrp):
-    if not vrrp:
+def verify(ha):
+    if not ha:
         return None
 
     used_vrid_if = []
-    if 'group' in vrrp:
-        for group, group_config in vrrp['group'].items():
+    if 'group' in ha:
```
Comment (c-po) [inline src/conf_mode/high-availability.py]: Shouldnt this be: if 'vrrp' in ha and 'group' in ha['vrrp']:?

File: smoketest/scripts/cli/test_ha_virtual_server.py (hunk 1 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_ha_virtual_server.py]: Replaced with `read_file(KEEPALIVED_CONF)`

File: src/conf_mode/high-availability.py (hunk 7 above)
Comment (sever-sever) [reply src/conf_mode/high-availability.py]: > Shouldnt this be: if 'vrrp' in ha and 'group' in ha['vrrp']:?

Sure, fixed
